torch/distributed/launch.py

```diff
@@ -198,6 +158,27 @@ def parse_args():
     parser.add_argument('training_script_args', nargs=REMAINDER)
     return parser.parse_args()
 
+
+def pool_processes(processes, cmd):
+    while 1:
+        exit_status = []
+        for process in processes:
+            return_code = process.poll()
+            exit_status.append(return_code)
+        # 1. all suceess, exit 0
+        # 2. one fail: terminal others, exit 1
+        # 3. filter suceess ,other is running, waiting
+        if all(map(lambda x: x == 0, exit_status)):
+            exit(0)
+        elif any(map(lambda x: x is not None and x != 0, exit_status)):
+            for process, return_code in zip(processes, exit_status):
+                if return_code is None:
+                    process.terminate()
```

Comment:
`process.terminate()` doesn't block until the pid disappears from the os, you also have to call `process.wait()` to ensure that that the subprocess has actually terminated before exiting the launcher process otherwise the subprocesses may become defunct or orphaned.